DRA scheduler: fix queuing hint support

pohly opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

d66f8f9 added that "plugins have to implement a QueueingHint for Pod/Update event if the rejection from them could be resolved by updating unscheduled Pods itself".

This applies to DRA because the name of a generated ResourceClaim must be recorded in the pod status before the pod can be scheduled.

#### Does this PR introduce a user-facing change?
```release-note
DRA: when enabling the scheduler queuing hint feature, pods got stuck as unschedulable for a while unnecessarily because recording the name of the generated ResourceClaim did not trigger scheduling.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/4381
- [KEP]: https://github.com/kubernetes/enhancements/issues/4247
```

/assign @sanposhiho

pohly committed:
DRA scheduler: fix queuing hint support

d66f8f9 added that "plugins have to implement a QueueingHint for Pod/Update
event if the rejection from them could be resolved by updating unscheduled
Pods itself".

This applies to DRA because the name of a generated ResourceClaim must be
recorded in the pod status before the pod can be scheduled.

sanposhiho commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -399,6 +399,8 @@ func (pl *dynamicResources) EventsToRegister(_ context.Context) ([]framework.Clu
 		// Allocation is tracked in ResourceClaims, so any changes may make the pods schedulable.
 		{Event: framework.ClusterEvent{Resource: framework.ResourceClaim, ActionType: framework.Add | framework.Update}, QueueingHintFn: pl.isSchedulableAfterClaimChange},
+		// Adding the ResourceClaim name to the pod status makes pods waiting for their ResourceClaim schedulable.
+		{Event: framework.ClusterEvent{Resource: framework.Pod, ActionType: framework.Update}, QueueingHintFn: pl.isSchedulableAfterPodChange},
```
In order to reduce the number of subscribing events, we avoid using `Pod/Update` and instead use [specific events `Pod/UpdatePodXXXX`](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/types.go#L69-L76).

So, I think you need to create a new one like `Pod/UpdatePodResourceClaims`.

alculquicondor replied:
I was going to ask whether adding a specific event is rather overkill, given that there is only one plugin that cares about it, and it is alpha.
In other words, we could hide the complexity inside the hint, as opposed to make it part of the framework.

But, it's all about prons and cons. I'll leave the decision to @sanposhiho

alculquicondor replied:
IIRC, the idea of fine-grained events is more important when there are no scheduling hints. But DRA is as close to graduating to beta as scheduling hints are, so maybe it's worth not adding extra complexity in the framework.

alculquicondor replied:
Oh right, having fine grained events is helpful to reduce the number of stored inflight events, which affects scheduling throughput even if DRA is not used in the cluster. That sounds worse for the average user. Ok, let's keep the fine-grained event :)

pohly replied:
Yes, I think that was the reason why @sanposhiho suggested it.

If DRA is disabled, the new event isn't getting generated. But if it's enabled (not on by default while beta!) and merely not used (no ResourceClaims), then it will add some overhead.

sanposhiho commented:
/cc @macsko

pohly commented:
/retest

pohly committed:
DRA scheduler: add special ActionType for ResourceClaim changes

Having a dedicated ActionType which only gets used when the scheduler itself
already detects some change in the list of generated ResourceClaims of a pod
avoids calling the DRA plugin for unrelated Pod changes.

googs1025 commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -399,6 +399,8 @@ func (pl *dynamicResources) EventsToRegister(_ context.Context) ([]framework.Clu
+		// Adding the ResourceClaim name to the pod status makes pods waiting for their ResourceClaim schedulable.
+		{Event: framework.ClusterEvent{Resource: framework.Pod, ActionType: framework.UpdatePodGeneratedResourceClaim}, QueueingHintFn: pl.isSchedulableAfterPodChange},
 		// A pod might be waiting for a class to get created or modified.
 		{Event: framework.ClusterEvent{Resource: framework.DeviceClass, ActionType: framework.Add | framework.Update}},
```
Does DeviceClass need the same operation? If so, will you submit it in this PR? Or can I submit another PR to complete this task and add some unit tests? @pohly

pohly replied:
I think DeviceClass is fine the way it is. It's very unlikely to change, so there is no risk of having to process too many events. That is the concern for pods.

googs1025 replied:
got it

toVersus commented on pkg/scheduler/framework/events.go:
```diff
@@ -222,6 +227,14 @@ func extractPodSchedulingGateEliminatedChange(newPod *v1.Pod, oldPod *v1.Pod) *C
 }
 
+func extractPodGeneratedResourceClaimChange(newPod *v1.Pod, oldPod *v1.Pod) *ClusterEvent {
+	if len(newPod.Status.ResourceClaimStatuses) != len(oldPod.Status.ResourceClaimStatuses) {
```
Since PodResourceClaimStatus doesn't have fields that are updated later, like state or timestamps, and entries added to PodResourceClaimStatus are never removed, does that mean we only need to compare the length of the slice? I saw the comment in another part of the code that caught my eye.
https://github.com/kubernetes/kubernetes/blob/52095a8b7b9b75d67a3882a21a6647e4f90ade48/plugin/pkg/auth/authorizer/node/graph_populator.go#L114-L116

pohly replied:
If we want to be pedantic, then yes, we would have to do a deep equal. As I decided to be pedantic elsewhere, let's to the same here :grin: 

Thanks for reminding me.

sanposhiho replied:
Can you leave the similar comment about why we only need check len?

pohly replied:
I'm now using a common helper function to check not only the length, but also the content.

sanposhiho commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -496,6 +501,33 @@ func (pl *dynamicResources) isSchedulableAfterClaimChange(logger klog.Logger, po
+		return framework.QueueSkip, nil
+	}
+
+	logger.V(4).Info("pod got updated and is schedulable", "pod", klog.KObj(pod))
```
Can you change the level to 5, if you don't have a strong preference. (Other QHints have level5 only

pohly replied:
Changed.

pohly replied:
Also in other, older code.

sanposhiho commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -496,6 +501,33 @@ func (pl *dynamicResources) isSchedulableAfterClaimChange(logger klog.Logger, po
+		// This is not an unexpected error: we know that
+		// foreachPodResourceClaim only returns errors for "not
+		// schedulable".
+		logger.Info("pod is not schedulable after being updated", "pod", klog.KObj(pod))
```
log level?

pohly replied:
Oops, yes. This was copy-and-pasted from above where it uses `loggerV` and then I forgot to add `V(6)` back.

sanposhiho reviewed (commented):
only nits

pohly committed:
DRA scheduler: reduce verbosity of queuing hints

Other hints also only use V(5) or higher.

sanposhiho commented on staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod.go:
```diff
@@ -0,0 +1,46 @@
+	// In most cases, status entries only get added once and not modified.
+	// But this cannot be guaranteed, so for the sake of correctness in all
+	// cases this code here has to check.
+	for i := range statusA {
```
Could statusA and statusB have the same items, but with different orders?

pohly replied:
In theory, yes. In practice that should never happen, so adding a `slices.Sort` here just adds performance overhead in the common case that doesn't pay off.

sanposhiho reviewed (approved):
/approve

Leave `/lgtm` to someone else, @macsko or @alculquicondor.
/cc @macsko @alculquicondor

pohly commented:
> pull-kubernetes-kind-dra: cluster supports count/resourceclaims.resource.k8s.io ResourceQuota 

A flake: the apiserver should prevent adding a second ResourceClaim, but does not always do so. A timing issue?!

>  pull-kubernetes-dependencies

@liggitt: this refactoring will need your approval because I am moving some helper function out of `authorizer/node` into a shared package.

But before this can be merged: what should I do about the [unwanted dependency on go-fuzz](https://github.com/kubernetes/kubernetes/pull/125487)? Are there alternatives that can be used for generating random test cases?

As I said in staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod_test.go: I could manually write test cases, but then the test won't catch when someone forgets to extend the code when adding some new field.

alculquicondor commented:
cc @dom4ha

liggitt commented on staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod.go:
```diff
@@ -0,0 +1,46 @@
+
+// PodStatusDeepEqual is a manually written DeepEqual for
+// [corev1.PodResourceClaimStatus] slices.
+func PodStatusDeepEqual(statusA, statusB []corev1.PodResourceClaimStatus) bool {
```
Does this treat `nil` and `[]` as equal? That's fine for semantic equivalence (and was fine for the authorization use), but could mess with callers who expect reflect.DeepEqual to agree with the results here.

pohly replied:
Semantic equivalence is sufficient.

pohly replied:
I'm comparing against reflect.DeepEqual, but apparently don't hit that particular kind of "not quite equal" when using go-fuzz.

liggitt replied:
> Semantic equivalence is sufficient.

I agree, mostly something we should make the naming / documentation clear on

pohly replied:
Renamed to just `PodStatusEqual` and updated the documentation.

I stopped using go-fuzz and replaced it with manually written testcases. That should be okay, no changes to `corev1.PodResourceClaimStatus` are expected. I'm still wondering what the replacement will be in other places were we use it.

pohly replied:
/assign @liggitt 

For final approval.

macsko commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -492,7 +497,34 @@ func (pl *dynamicResources) isSchedulableAfterClaimChange(logger klog.Logger, po
+// isSchedulableAfterPodChange is invoked for update pod events reported by
+// an informer. It checks whether that change adds the ResourceClaim(s) that the
+// pod has been waiting for.
+func (pl *dynamicResources) isSchedulableAfterPodChange(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) (framework.QueueingHint, error) {
```
Don't we need any tests for that?

googs1025 replied:
hey, @macsko @pohly I am working on this PR, which seems to be related to this. If necessary, I can add this part of the test case to this PR :)
https://github.com/kubernetes/kubernetes/pull/127523

pohly replied:
@macsko: yes, this needs tests. Is it okay if @googs1025 adds them as a follow-up?

macsko replied:
Sure, can be this way

alculquicondor replied:
This is not how we do things. And even more so when we have gotten rid of "test freeze", meaning that code needs to merge with all appropriate coverage.

I've let similar situations slide in the past, but IIRC, DRA is targeting beta now, so we can't be as forgiving as we were in previous iterations of DRA.

johnbelamaric replied:
@alculquicondor +1, we need to be sure we merge code and tests together from now on

pohly replied:
Fair enough. I've added unit tests.

macsko commented:
/lgtm

dom4ha commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -492,7 +497,34 @@ func (pl *dynamicResources) isSchedulableAfterClaimChange(logger klog.Logger, po
+		// schedulable".
+		logger.V(6).Info("pod is not schedulable after being updated", "pod", klog.KObj(pod))
+		return framework.QueueSkip, nil
+	}
```
The alternative would be to just check whether the generated resource claims has shrunk down and enable scheduling, but I suspect this was a conscious decision.

dom4ha replied:
Actually shrunk down or changed, as in theory they claims can change as well. So not sure what is the most common case and whether the additional complication is worth the effort, but changes that are append only, won't infer the scheduling I think.

pohly replied:
This hint function only gets called if the code over in `events.go` has already determined that the pod's generated claims changed, so there is no need to check that again.

What the code here checks is "are all required claims generated".

dom4ha replied:
> This hint function only gets called if the code over in `events.go` has already determined that the pod's generated claims changed, so there is no need to check that again.

For instance instead of introducing a generic change event UpdatePodGeneratedResourceClaim, the alternative one could be scoped down (emitted) to changes that shank down the Pod's claims (if any of them disappeared - PodGeneratedResourceClaimEliminated) as this is the only case worth requeuing.

Unfortunately I don't know how often it happens that new ResourceClaims just grow, as this is the only case this approach could eliminate (isSchedulableAfterPodChange would not be called).

pohly replied:
> the alternative one could be scoped down (emitted) to changes that shank down the Pod's claims (if any of them disappeared - PodGeneratedResourceClaimEliminated) as this is the only case worth requeuing

It's the opposite: *adding* entries may make a pod schedulable because the ResourceClaim that it needs got added.

In practice, entries only get added, never removed nor updated. I don't think it's worth making the code more complex for situations that don't occur.

dom4ha replied:
Ah right, claims gets fulfilled over time. There's indeed no place for optimization.
Sorry for the dumb question, just ramping up and connecting dots slowly :)

dom4ha commented:
LGTM

alculquicondor commented on staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod.go:
```diff
@@ -0,0 +1,52 @@
+		if statusA[i].Name != statusB[i].Name {
+			return false
+		}
+		claimNameA := statusA[i].ResourceClaimName
+		claimNameB := statusB[i].ResourceClaimName
+		if (claimNameA == nil) != (claimNameB == nil) {
+			return false
+		}
+		if claimNameA != nil && *claimNameA != *claimNameB {
+			return false
+		}
```
I think you can use `ptr.Equal`

pohly replied:
Indeed, thanks for the tip.

pohly commented on test/integration/scheduler_perf/.gitignore:
```diff
@@ -1 +1,2 @@
 BenchmarkPerfScheduling_*.json
+
```
Done.

pohly commented:
/retest

alculquicondor commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:
```diff
@@ -1497,6 +1502,67 @@ func Test_isSchedulableAfterClaimChange(t *testing.T) {
+			expectedErr: true,
+		},
+		"complete": {
+			objs:         []apiruntime.Object{pendingClaim},
```
we don't check these objects in the function, do we?

pohly replied:
We do, indirectly in https://github.com/kubernetes/kubernetes/blob/3d6c5b2e98afaaae1d17107e2d3d709c726be49d/pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go#L683-L686

As part of my work on https://github.com/kubernetes/kubernetes/pull/127277 I checked whether removing it would make a difference, but couldn't measure any, so I decided to keep the `Get` as an additional check.

alculquicondor commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:
```diff
@@ -1497,6 +1502,67 @@ func Test_isSchedulableAfterClaimChange(t *testing.T) {
+		obj          interface{}
+		expectedHint framework.QueueingHint
+		expectedErr  bool
+	}{
```
add a case for different UID

pohly replied:
Added.

alculquicondor commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:
```diff
@@ -1497,6 +1502,67 @@ func Test_isSchedulableAfterClaimChange(t *testing.T) {
+				EnableDynamicResourceAllocation: true,
+			}
+			testCtx := setup(t, nil, tc.claims, nil, nil, tc.objs, features)
+			actualHint, err := testCtx.p.isSchedulableAfterPodChange(logger, tc.pod, nil, tc.obj)
```
```suggestion
			gotHint, err := testCtx.p.isSchedulableAfterPodChange(logger, tc.pod, nil, tc.obj)
```

pohly replied:
That's not what is used elsewhere, for example:
https://github.com/kubernetes/kubernetes/blob/3d6c5b2e98afaaae1d17107e2d3d709c726be49d/pkg/scheduler/framework/plugins/nodeports/node_ports_test.go#L305-L358

But I don't care, so I changed it in `dynamicresources_test.go` also in the other functions.

alculquicondor commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:
```diff
@@ -1497,6 +1502,67 @@ func Test_isSchedulableAfterClaimChange(t *testing.T) {
+		pod          *v1.Pod
+		claims       []*resourceapi.ResourceClaim
+		obj          interface{}
+		expectedHint framework.QueueingHint
+		expectedErr  bool
```
```suggestion
		wantHint framework.QueueingHint
		wantErr  bool
```

alculquicondor commented on staging/src/k8s.io/dynamic-resource-allocation/resourceclaim/pod_test.go:
```diff
@@ -0,0 +1,78 @@
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
```
In SIG scheduling we prefer not to use assertion libraries. Please use `t.Error`

alculquicondor replied:
https://github.com/kubernetes/community/blob/master/sig-scheduling/CONTRIBUTING.md#technical-and-style-guidelines

pohly replied:
As SIG Testing tech lead I don't agree with those guidelines. I accept it for code owned by SIG scheduling, but this here is a general-purpose helper package.

liggitt replied:
Testify keeps us tied to a yaml library we want to drop eventually (https://github.com/stretchr/testify/blob/master/go.mod#L11), it would be vaguely nice to not add more deps to it

liggitt replied:
(But not a blocker, certainly)

pohly replied:
https://github.com/stretchr/testify/pull/1579 added build tags which avoid that dependency. I'll have a look whether we can leverage that.

liggitt replied:
We can't :-(

go dependency and vendoring don't consider build tags since they have to resolve ahead of when builds are done and be valid for whatever builds *could* be done

alculquicondor commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:
```diff
@@ -1497,6 +1502,67 @@ func Test_isSchedulableAfterClaimChange(t *testing.T) {
+				return
+			}
+
+			require.NoError(t, err)
```
Avoid assertion libraries and write meaningful errors for `t.Error`

pohly replied:
Okay, changed it here.

liggitt commented:
/approve
For authorizer change

alculquicondor reviewed (commented):
/lgtm
/approve
